### DISCUSSION TITLE
Improving HC's atomic execution protocol

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#154] [General]
As brought up by @aakoshh [here](https://github.com/adlrocha/builtin-actors/pull/13#issuecomment-1195858240), the current implementation of the atomic execution protocol doesn't include any checks on the locked states, allowing users to use the same locked state in several executions without anyone realizing it. 

[Here](https://github.com/adlrocha/builtin-actors/pull/13#issuecomment-1195879040) is @aakoshh's proposal to improve the protocol.

> * Participants call `lock` on some state of theirs to get the corresponding input CID.
> 
> * They all agree on the `AtomicExecutionParams` from the `msgs` and `inputs` they now know.
> 
> * One of them sends the `AtomicExecutionParams` to the nearest common ancestor, they all learn the resulting CID of the record - although it's a good question how they can trust anyone telling them what it is because of changing the address strings to IDs. Maybe the CID should be that of the original, raw params?
> 
> * _New step:_ Everyone calls `pre_commit(lock_cid, atomic_exec_params_cid, output)` _in their own subnet_. The lockable actor now associates the locked data with this execution, and sends the commitment to `output_cid` up the chain as a bottom-up message.
> 
> * The SCA in the atomic execution subnet gathers to pre-commits, and issues the `commit(atomic_exec_params_cid)` once everyone's vote has arrived.
> 
> * The actors in the descendant subnets receive the commit message and do the unlock.

The goal of this discussion is to brainstorm ways to strengthen the protocol for its first version. Also worth considering this [potential attack-vector](https://github.com/filecoin-project/eudico/issues/228) in the discussion. The outcome of the discussion should result in an update of the spec with the corresponding changes.

------
__Edit__: Adding a sample use case where several types are used for an atomic execution (shared by @aakoshh in Slack):
> The following illustrative example came to my mind to demonstrate how types in the atomic execution protocol actors can be different. It's about horse racing.Say we have the following three subnets:
>
> 
> -   Stadiums: venues where people can place bets and the money is handled
> -   Stables: where horses are bred, trained and kept
> -   Jockeys: people who can ride horses, but don't own them
> 
> A race is an atomic execution function which requires 3 inputs:
> 
> -   A stadium to organise the race, taking some fee and distributing people's winnings
> -   Some horses from some stables, depending on availability
> -   Some jockeys to ride them, if they are not on holiday or doing other races
> 
> The output of the race is the ranking of the fastest horse and jockey pairings, from which everyone can calculate if, and how much they won (the jockey and the stables split the reward).We can see that everything has a different type:
> 
> -   Horses don't have the same type as the stable they came from
> -   Horses don't have the same type as the jockey that sits on them
> -   The output doesn't share a type with anything, it's enough if any subnet can look at it and decide how to update their own state.
> -   All inputs are different type, but that's fine because the execution is off-chain and the participants can pull the individual inputs separately and pass them to the execution as the correct type, no need for the merge function in this case.
> 
> On a more practical note, what I'm not sure about is how the circulating supply of the subnets can be kept up to date. I think the pre-commit I proposed can calculate from the output how much funds it has to release up to the common subnet, or indeed how much it is owed. Once everyone has done their pre-commits and they reach the SCA, the SCA has to redistribute the funds according to the agreement and send the correct values back in the top-down commit messages, increasing the circulating supplies where credits are due.So the stadiums would release some money towards the SCA, then the stables and the jockeys would get it. The rest of the money is distributed to punters within the stadium, they don't leave that subnet.It is important that released money is returned if there is an abort, the pre-commit cannot get lost.

**Comment by adlrocha** [comment]
As suggested in the proposal, we need a way to link the locked state in the source subnet with the specific execution. I think a modified version of @aakoshh proposal could work. Leaving a few comments/questions in-line.

> * Participants call `lock` on some state of theirs to get the corresponding input CID.
> 
> * They all agree on the `AtomicExecutionParams` from the `msgs` and `inputs` they now know.
> 
> * One of them sends the `AtomicExecutionParams` to the nearest common ancestor, they all learn the resulting CID of the record - although it's a good question how they can trust anyone telling them what it is because of changing the address strings to IDs. Maybe the CID should be that of the original, raw params?

I think I left a comment somewhere in the code, but yes, the translation into IDs is for internal use in the SCA, to avoid having to resolve the signable addresses every time, but the CID of the execution should be using the raw params (i.e. including the original signable addresses of the parties involved).

> * _New step:_ Everyone calls `pre_commit(lock_cid, atomic_exec_params_cid, output)` _in their own subnet_. The lockable actor now associates the locked data with this execution, and sends the commitment to `output_cid` up the chain as a bottom-up message.

Why a bottom-up message? I was thinking that we could lock the state in the subnet (without linking it to any execution), then when the execution is initialized, the owner needs to link it to the cid of the execution, and to propagate the output the user sends a message directly to the comon parent (without requiring a bottom-up message) but including the `pre_commit` information that allows the SCA to link the locked state to the execution. 

> * The SCA in the atomic execution subnet gathers to pre-commits, and issues the `commit(atomic_exec_params_cid)` once everyone's vote has arrived.
> * The actors in the descendant subnets receive the commit message and do the unlock.
>
> So, the change are that the output isn't coming from a top-down message, it's the participants entering it into their respective subnets; doing so triggers a pre-commit message going up, and the locked state being associated with the atomic execution, so it can't be associated with anything else, until abort or commit arrives.

I don't think I understand this last paragraph. Would you mind elaborating a little? Thanks :pray:

**Comment by aakoshh** [comment]
Thanks for creating this issue @adlrocha ! 

> but the CID of the execution should be using the raw params

Thanks, I saw your comment after I sent this one. Cool, in that case the CID can be known up front. I will ask about the ID in the original PR.

> I was thinking that we could lock the state in the subnet (without linking it to any execution), then when the execution is initialized, the owner needs to link it to the cid of the execution

Yes! My suggestion is exactly to 1) lock 2) init then 3) link to CID as a "pre-commit" step. 

I think the "link it to the cid of the execution" part is missing from the [current spec](https://github.com/protocol/ConsensusLab/blob/hc/spec/specs/hierarchical_consensus.md#implementation-of-the-protocol). In the spec, you 1) lock state with `(method, params)`, then 2) go to the SCA to init. I think `(method, params)` does not identify the atomic execution uniquely (there's nothing to suggest that it does, and I can give examples of swaps where I think the `params` are different in each subnet). After this there's 3) off-chain execution and 4) sending the `output` to the SCA which will 5) send a top-down message with `(method, params, output)` to each participant.

Because `params` is not unique, you can't be sure whether somebody has not unlocked their state already by using the locked state in two different atomic executions. If they were linked by the atomic execution CID, this is no longer a problem.

> Why a bottom-up message? 

The pre-commit step I propose links the locked state to the atomic execution at the same time. Now, only a bottom-up message can be trusted to actually tell the SCA that this linking has happened. Any user participating in the atomic execution could just lie about it, in which case some participants do their bit of the transaction, and the liar does not. By using a bottom-up message, we're not trusting the user, we're trusting the validators of the subnet and the LockableActor, which emitted the message _only_ after the linking has actually happened. Checkpoints are involved, so there's no going back from it either.

> I don't think I understand this last paragraph. Would you mind elaborating a little? 

Sorry, that was dense. 

The first half was about how the actual `output` enters the chain. In the spec and the PR the `output` is submitted as raw bytes to the SCA as many times as there are participants, then it travels down as a top-down messages to each subnet, through each intermediary. The `output` is actually known to the participants, they calculated it; the only reason `output` has to travel into the participants' subnets is because otherwise it could be a data availability attack: the actor could not merge it into its state, if the participant does not enter it, a CID is not enough. But, if we allow the participant to add the `output` to its own subnet up-front during pre-commit, it doesn't have to travel through all the other intermediary subnets, only its CID goes up, and then the commit/abort comes down. It involves less moving data across the hierarchy, but doubles the number of (smaller) messages.

The second half just reiterates that `output` is sent as part of the pre-commit, along with the atomic execution ID, to be associated with the locked state, until this atomic execution ID is committed or aborted, at which point it can be merged, as it's already present in the ledger.

**Comment by aakoshh** [comment]
There is a timing issue with the proposal that we have to be aware of.

We agreed in the PR that the SCA doesn't have to remember the atomic executions that have concluded. It is possible that Alice and Bob agree on the parameters of an atomic execution, then the following event sequence occurs:
1. Alice registers the atomic execution in the SCA. 
2. Alice links the atomic execution to her state. 
3. Alice aborts the atomic execution, the SCA sends the top-down message and forgets the atomic execution.
4. Alice unlocks her state.
5. Bob hasn't linked his state so he discards the abort message.
6. Bob links the atomic execution to his state (unaware of the abort).
7. Bob does the pre-commit and sends the bottom-up message.
8. The SCA doesn't remember the atomic execution and discards the message (it might go all the way up to the root, depending on how it's implemented).

Now, Bob could be sitting there for a while. To unlock his state, he would have to re-initiate the same atomic execution with the SCA and abort it.

Maybe if the SCA doesn't recognise the atomic execution in the pre-commit it could automatically send an abort. Or just keep the tombstone around for a while.

**Reply by adlrocha** [reply]
I think this timing issue can be easily fixed. Locked state can only be used for a single execution, right? So even if Bob hasn't linked his state yet, if we still use a top-down message from the parent to propagate the unlock message triggered by the abort, when it arrives to the subnet, it will check the locked state affected by the abort, and will unlock it even if it hasn't been linked yet (I think the timing issue is easier to perform when the user is responsible for unlocking the state, instead of the SCA through a top-down message). 

(I'll add a thread below putting all that we've discussed together to see if we agree on the proposal :) )

**Reply by aakoshh** [reply]
I don't know, when the top-down message is coming, it might just say that it aborted a specific atomic execution CID. If you haven't linked, you don't know which input it refers to. 

I keep thinking about this, to make sure no funds get lost if they are carried by bottom-up and top-down messages. 

The obviously correct solution is to keep the tombstone of the atomic execution in the SCA forever, and then it can respond to a pre-commit any time it comes, returning the funds that it comes with. I reckon we should start with that because it's the easiest to get right. 

Another timing issue I see is what if I pre-commit before the atomic execution is actually initialised in the SCA. Where do my funds end up going? The problem here is: what should be in the pre-commit so that the SCA in the right common parent recognises that it should be the one executing it, but it doesn't have any record of it, so it should immediately return the funds. 

Or alternatively the pre-commit should contain so much information that it can serve as the init itself, so any pre-commit arriving from any subnet is the init, and there's no outside user gesture to do the init outside the subnet comm system. That way if Alice aborts first, then Bob's pre-commit will just re-establish the atomic execution, which he can then abort. The problem is that we would never be quite sure what happened to the atomic execution: it was aborted, now it's back on? And none of this is a problem if we don't delete the tombstone. 

Again the important thing is for the circulating supplies to be maintained correctly in all even sequences. 

Timeouts could be applied, so the subnet won't even send a pre-commit if it's too late, and the SCA keeps the tombstone long enough for every subnet to be able to propagate all the checkpoints. But it's difficult to say what they should be to make sure this is true.

**Reply by adlrocha** [reply]
>I don't know, when the top-down message is coming, it might just say that it aborted a specific atomic execution CID. If you haven't linked, you don't know which input it refers to.

But if you provide the input params in the abort message, you know what state to remove even if it's not linked, right? Because there's no way you've been able to unlock it and use for other executions.

> The obviously correct solution is to keep the tombstone of the atomic execution in the SCA forever, and then it can respond to a pre-commit any time it comes, returning the funds that it comes with. I reckon we should start with that because it's the easiest to get right.

Although maybe it's a good idea to do this initially just to be safe.

> Or alternatively the pre-commit should contain so much information that it can serve as the init itself, so any pre-commit arriving from any subnet is the init, and there's no outside user gesture to do the init outside the subnet comm system. That way if Alice aborts first, then Bob's pre-commit will just re-establish the atomic execution, which he can then abort. The problem is that we would never be quite sure what happened to the atomic execution: it was aborted, now it's back on? And none of this is a problem if we don't delete the tombstone.

I like this approach. If we keep the tombstone we can immediately return the funds if it was already initialized.

**Comment by adlrocha** [comment]
Summary of the proposal for the improved atomic execution protocol:

- Participants call `lock` on some state of theirs to get the corresponding CID of the state to be use as input in the execution.

- Participants agree on the `AtomicExecutionParams`, `msgs` and `inputs` used for the execution.

- One of the parties initializes the execution by sending the `AtomicExecutionParams` to the nearest common ancestor. They all learn the CID that uniquely identifies the execution (and that can deterministically be computed from the `AtomicExecParams`).

- All parties collect the input CIDs from the other subnets involved and perform the execution off-chain to get the output state.

- The submission of the output is done in a new step that we call the `pre_commit` stage. All lockable actors include a new `pre_commit(lock_cid, exec_cid, output_cid)` function that receives as input the cid of the locked inputs, the execution cid, and the output cid. This function links its own state the locked state and the output state to the current execution (so it can't be used as input for any other), and triggers a bottom-up transactions propagating the tuple `(exec_cid, output_cid, cid(lock_cid, exec_cid))` to the common ancestor to submit the execution. When the message arrives to the SCA, it checks that for the execution being submitted, the `output_cid` maps those of the other parties, and that the cid of the `lock_cid` of the inputs provided and the `exec_cid` match the ones sent in the initialization stage (thus verifying the link between input states with the execution).

- Once the `pre_commit` messages of all participants have reached the common ancestor, the SCA triggers a top-down message signaling that the execution succeeded and that the output state can be merged and the locked states unlocked.

- If half-way any of the parties want to abort the execution, they just need to send an `abort` message directly to the parent of the common ancestor. This will trigger a top-down `abort` message to unlock the state in all subnets.

This proposal removes the issue of a user using the same locked state for several executions, and by submitting the result of the execution through a bottom-up message instead of a direct message by the party involved, we put the trust on the corresponding subnet's validators (the message needs to be verified to be propagated up) and not on a single user.

@aakoshh can you confirm that this is in-line with what you were thinking?

**Reply by aakoshh** [reply]
Thanks for the summary. Some differences:
* `pre_commit` should take `output`, with the `output_cid` computed by the lockable actor, which also stashes the `output` for later. In your summary above the `output` never entered the system, so the actor could be blocked trying to resolve it.
* I don't see why we should send `cid(lock_cid, exec_cid)` instead of just `lock_cid`. I agree though that including proof that the agreed up on input was used is good! 

> triggers a bottom-up ... to the common ancestor to submit the execution

It's not clear how the `pre_commit` would know who the common ancestor is. We can say it can go up until _someone_ recognises the `exec_cid`, but as per my other comment, it's another race condition. 

I'm starting to think that the most bomb proof protocol would be:

1. Participants `lock` some state for input and get the CID for `inputs`.
2. Participants agree on the `AtomicExecutionParams`, which includes the `inputs`. 
3. _Change_: nobody talks to the SCA to init anything. 
4. Participants collect inputs and do the offline execution to compute the `output`. 
5. Participants call `pre_commit(AtomicExecutionParams, output)` in their subnet: the lockable actor can look at the params and see what `lock_id`(s) from the current subnet are associated with this execution, and link them, unless the lock is already linked to something or it's unlocked. The lockable actor can also compute the common ancestor and knows where to send the bottom-up message. It sends `(exec_cid, output_cid, debit_credit_map)` to the common ancestor, and releases the funds if it's debited. The debit-credit mapping should be computable from the `output`. The `lock_cid` is not part of the message here because this time `exec_cid` is computed by the lockable actor itself, so any discrepancy in the `lock_cid`s would lead to the pre-commit failing, or having a different `exec_cid`.
6. Once the messages reach the SCA, it checks that everyone agrees on the `debit_credit_map` and the `output_cid` and sent the right amount of funds, then sends top-down messages `commit(exec_cid)` attaching the funds due to each subnet, or `abort(exec_cid)` and a refund if there's a disagreement. 

If someone wants to abort, they can do so once the atomic execution is established at the common ancestor by a pre-commit. The SCA should keep a tombstone and return funds immediately for later pre-commit messages. 

Parties can also decide not to call pre-commit in their own subnet, in which case they should be allowed to `unlock` their inputs and go on with their lives. The others might do the same, or they might pre-commit and abort at the common ancestor. 

So, to summarize: I propose now that we exclusively communicate to our respective subnets the whole parameter set, to avoid all timing issues.

**Reply by adlrocha** [reply]
>pre_commit should take output, with the output_cid computed by the lockable actor, which also stashes the output for later. In your summary above the output never entered the system, so the actor could be blocked trying to resolve it.

Yes, this is what I meant (maybe I didn't describe it correctly). 

> We can say it can go up until someone recognises the exec_cid, but as per my other comment, it's another race condition.

I think this works.

Overall I am happy with how this looks like :)

**Comment by LefKok** [comment]
I am adding some resources on replay attack defences on cross-shard transactions. TL;DR when running a cross-net messages two things need be done to protect against replay attacks:
1) Every object needs to have a nonce that increments, this creates a session identifier to avoid replaying
2) Receiving/Output subnets of a transaction (i.e., the once that create new assets) need to check that they create new assets once. This can be done either but checking the chain for uniqueness (which can be costly) or by having some state that initializes a dummy asset (with value 0) and then spends it to a real asset when the communication happens. 

Resources:
Paper:https://ieeexplore.ieee.org/abstract/document/9230373 
Video:https://www.youtube.com/watch?v=5qpgarMYW70 (first half)

**Comment by sergefdrv** [comment]
I think there is another caveat left in the [last proposed version of the protocol](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3267610), namely the output state being computed off-chain is not subject to the same correctness guarantees as our users might expect. 

I can't see correctness of the output state checked anywhere on-chain. If all the users involved in an atomic execution protocol collude and pre-commit to an incorrectly computed output state to their advantage, the fabricated output state would get merged into the involved lockable actors' state. Whereas in some use-cases, it will not be the most rational behaviour, in general, this may result in beneficial outcomes for the adversarial users. Therefore we should account for that. 

I think we really need to compute the output state on-chain such that its correctness is ensured by the underlying consensus protocol, as one would reasonably expect.

**Reply by aakoshh** [reply]
I thought this was a feature, not a bug. If you put the execution on-chain then every validator in the network has to run the computation, whereas if it's off chain then it's only the participants that need to dedicate resources, and they can use any technology they want, not limited by what they can deploy on chain. 

You are right that this means they can do anything they can agree on. The limitation on the damage is the that the messages have to be accompanies by a _value_ that increases/decreases the circulating supply in the subnet they go to/from, and that value has to be somehow determined by the actor.

Maybe another way would be to use Lurk and ZK proofs for correctness. But this is like a side channel: as long as no new funds are created, it should be up to the participants to agree on how they swap around their stuff, no?

**Reply by aakoshh** [reply]
We never actually put the inputs on the chain to do the computation, IIRC the `AtomicExecutionParams` doesn't contain data, just CIDs of it, so it's light weight. Who knows how much data it would require to put all other inputs from all other subnets on every participating subnet to do the computation?

**Reply by sergefdrv** [reply]
From my conversation with @adlrocha, I got an impression that adjusting circulating supply of subnets should not be a part of the atomic execution protocol, since we are working with arbitrary pieces of state. @adlrocha could you please expand on this matter?

**Reply by sergefdrv** [reply]
To support use cases with intensive computation in [the scheme proposed below](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477), we can make `exec_actor.init_atomic_exec()` return some extra opaque data along with concise _input_. Then we can use the extra data to do whatever off-chain computation we wish and supply the result into `exec_actor.prep_atomic_exec()` as an additional argument along with the concise `inputs`. The rest should work the same. But I would argue it would be a safer default choice to compute the output on-chain because I believe the users would naturally expect the guarantees of on-chain execution for atomic executions.

**Reply by aakoshh** [reply]
>  I believe the users would naturally expect the guarantees of on-chain execution for atomic executions.

I understand where you are coming from, I'm just saying that both schemes have a right to exist, as they do with side channels. My impression of the atomic execution protocol was that it was the participating actors' responsibility to make sure the state is updated in a way they agree, and not everyone else's responsibility to police them. 

I'd add that I think it's also not impossible to do what you want with the original protocol. There are two finishing steps in that protocol that would allow this:
1. In the _pre-commit_ everyone submits the `output` to the actor in their respective subnets; this `output` has to be the same for everyone, but the protocol doesn't say what it is, it's up to the actors.
2. When the _commit_ message arrives from the common ancestor coordinating the atomic execution, everyone looks up the `output` and _selects_ the part relevant to their subnet, and _updates_ their state. 

There is nothing stopping someone from putting all the gathered inputs into the `output` and use the _update_ step to do the calculation on-chain that modifies the subnet state. This way it will be verified by every validator on the chain, and all the participants have to agree on are the inputs that go into the output.

**Reply by aakoshh** [reply]
> I got an impression that adjusting circulating supply of subnets should not be a part of the atomic execution protocol, since we are working with arbitrary pieces of state

I had a conversation about this with @guy-goren in Istanbul and I thought I manage to convince him otherwise, but now I'm not sure either. 

The example I had in mind at the time was that Alice has 10 apples in subnet A and wants to swap with Bob for 5 FIL in subnet B. It seems like if FIL is involved then it should decrease the circulating supply of B and increase A. But if we say that we want Bob to send 20 Bananas instead, we'd need a price oracle to tell what is the balance of change. My thought there was to let the participants supply the balance of change as part of the pre-commit, and let the actors validate that they have enough FIL to cover it if they are in debit. 

If this is not part of the protocol then it sounds like Bob has to first transfer his FIL to A and buy the apples there, then do an atomic execution which destroys the apples in A and mints them in B. In another example we could swap an invaluable NFT of the Mona Lisa in A to a bag of crisps in B with no adjustment of circulating supply, and if people want to buy the NFT in B then they have to bring in the FIL. 

Yeah, I'm happy with this, thanks :+1:

**Reply by adlrocha** [reply]
I am probably joining the discussion late and this doesn't add much to it, but my take is that for the initial "improved" implementation of the protocol: 
- I aggre with @aakoshh point, _"is the participating actors' responsibility to make sure the state is updated in a way they agree, and not everyone else's responsibility to police them."_. App developers that want to support atomic execution on their actors need to write the logic of their actor (through the implementation of `Lock`, `Merge`, `Unlock`, etc.) so that only the owner of some state is allowed to lock, merge, or use some state for an atomic execution. Let's take an actor that implements and NFT marketplace as an example: a user may lock their state for some NFT exchange and mess with it to merge some random NFT from other subnet after colluding with a user in some other subnet, but that is fine because if the actor is implemented correctly, the colluding parties can only mess with each others state (NFTs), that's it (of course, as long as the validators of the corresponding subnets are honest). 
- Regarding the update of the circulating supply, this value only represents the amount of FIL (i.e. native token) in the subnet, so as long as the atomic execution doesn´t  move FILs, there is no update required (we are not currently considering any TVL of the assets locked in state of a subnet) --IIUC what you were discussing--.
- Finally, without fraud proofs or some way of achieving off-chain verifiable computation, I don't think we can have super strong security guarantees (without executing the job on-chain, in which case we may be killing the purpose of having different subnets all along, because we would be pushing a lot of state for the on-chain execution to the common parent). Even more, I think we shouldn´t  even attempt to enforce the firewall requirement for state.  

TL;DR: I would say the current "improved" version of the protocol that @aakoshh came up with is "good enough" security and performance-wise. I agree that we could do better, but that would require either integrating ZK technology like Lurk, or pushing a lot of state and potential computation to the common parent. _(that being said, I haven´t read @sergefdrv's proposal below yet, I'll follow-up on that on the corresponding thread)_

**Reply by sergefdrv** [reply]
> > I got an impression that adjusting circulating supply of subnets should not be a part of the atomic execution protocol, since we are working with arbitrary pieces of state
> 
> I had a conversation about this with @guy-goren in Istanbul and I thought I manage to convince him otherwise, but now I'm not sure either.

I think we better first focus on the pure atomic execution protocol without adjusting subnets' circulating supply. Later we could add moving funds as a part of the protocol.

**Reply by sergefdrv** [reply]
> participating actors' responsibility to make sure the state is updated in a way they agree, and not everyone else's responsibility to police them

> pushing a lot of state and potential computation to the common parent

I'm not proposing to push any extra computation, responsibilities, or policies to the common parent.

**Comment by sergefdrv** [comment]
Following is my attempt to fix the issue with computing the output
state off-chain, as well as the replay attacks. The protocol is rather
generic, but I think we can easily instantiate it in a specific form
for the initial implementation.

The atomic execution protocol involves the following entities:

 - A fixed number of _execution actors_ which are responsible for
   carrying out the use-case-specific part of the atomic execution.
   Execution actors can be instantiated in different subnets, whereas
   each of them has a globally unique identifier - _actor ID_.
 - A single _coordinator actor_ which coordinates the atomic
   execution. The coordinator actor is instantiated in a common
   ancestor subnet for all the execution actors.
 - A number of _users_ authorized to invoke the execution actors as
   described by the atomic execution protocol.

**Initialization.** To initiate the protocol, the users invoke a
dedicated method on each of the execution actors:

    exec_actor.init_atomic_exec(...)
  
The method may receive some arguments and returns an opaque result.
The obtained result will be referred to as _atomic execution input_
and used in the next step of the protocol.

_Edit: The opaque result should include a_ concise input ID _to be
used at the next step of the protocol. The result may also include
some additional data, e.g. to be used in the additional argument at
the next step._

The execution actor must return _distinct_ results on each invocation,
e.g. by maintaining a monotonic counter which is incremented on each
invocation of the method and used to produce the result.

The actor may lock a portion of its state until the atomic execution
is complete, such that it rejects any attempt to modify or lock that
portion of the state.

**Commit.** The users exchange the atomic execution inputs obtained at
the initialization step and, once collected all the expected inputs,
invoke another dedicated method on each of the execution actors:

    exec_actor.prep_atomic_exec(inputs, ...)

The method receives a set of inputs indexed by the execution actor IDs
and some other optional arguments.

On a successful invocation, the actor must _consume_ its own input
such that any subsequent attempt to re-use that input is rejected.

The method deterministically computes a concise _atomic execution ID_
which must be _unique_ for each distinct set of inputs (including
actor IDs) and such that all the execution actors compute the same ID.

If not done at the previous step, the actor locks the portion of its
state affected by the atomic execution, such that it rejects any
attempt to modify or lock that portion of the state until the atomic
execution is complete.

The execution actor further computes the atomic execution output and
stores it in its state associated with the atomic execution ID.

The method emits a bottom-up message invoking a dedicated method of
the coordinator actor specifying the set of actor IDs, the computed
atomic execution ID, as well as its method identifies to call back for
committing or rolling-back the atomic execution:

    coord_actor.pre_commit(actors, exec_id, commit, rollback)

_Edit: The `rollback` parameter should be removed._

The coordinator actor records the invocation in a set associated with
the pair `(actors, exec_id)`. Once the set contains entries for all actor IDs, 
it emits a top-down message for each of the actors invoking their commit 
methods:

    exec_actor.commit(exec_id)

Upon invocation of the commit method, the execution actor completes
the atomic execution by releasing any state previously locked for the
atomic execution, removing from its state the output associated with
the atomic execution ID, and then applying the output to the state.

**Abort.** The user may simply cancel an initiated atomic execution
before the atomic execution input has been consumed in the execution
actor by invoking a dedicated method on the execution actor:

    exec_actor.cancel_atomic_exec(...)

The method receives an argument specifying the atomic execution input
produced at the initialization step. 

On a successful invocation, the actor _consumes_ the input such that
any subsequent attempt to re-use that input is rejected. It also
releases any state previously locked for the atomic execution.

After the commit step, the user can abort the atomic execution by
invoking a dedicated method on the execution actor:

    exec_actor.abort_atomic_exec(...)

The method receives arguments specifying the atomic execution that was
prepared, but not yet committed by the actor at the commit step.

The method simply emits a bottom-up message invoking a dedicated
method of the coordinator actor specifying the set of actor IDs and
the atomic execution ID, as well as a method identifier to call back
for rolling-back the atomic execution:

    coord_actor.abort(actors, exec_id, rollback)

The coordinator actors removes the set associated with the pair
`(actors, exec_id)` from its state and emits a top-down message for
each of the actors recorded in the set, as well as for the caller
(which must be among the actor IDs), invoking their rollback
methods:

    exec_actor.rollback(exec_id)

_Edit: The coordinator actor only removes the record corresponding to 
the execution actor which emitted the bottom-up message and sends 
a top-down message to that actor._

Upon invocation of the rollback method, the execution actor terminates
the atomic execution by discarding from its state any stored output
associated with the atomic execution ID. It also releases any state
previously locked for the atomic execution.

**Reply by sergefdrv** [reply]
We could also leave a tombstone in the coordinator actor state for aborted atomic executions such that any executor actor attempting to pre-commit that atomic execution will immediately receive back a top-down message invoking its rollback method.

**Reply by adlrocha** [reply]
The main difference between the [latest proposal](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3268374) of the protocol and what you propose here is that once the inputs for the execution are exchanged by all parties, every subnet executes on-chain on their subnet, and propagate the result up to the coordinator from the execution actor. The coordinator waits for all outputs before signalling the success of the execution, and allowing the final commit in the execution actor states, right? If there is an abort, on the other hand, the coordinator signals the execution actors to rollback to the previous state before locking. Did I understand the proposal correclty?

This would mean an additional RTT of cross-net messages from the execution actors and the coordinator to notify the output of the on-chain execution, but apart from that I guess it works in a similar way (I don't know if it fully removes the issue of users colluding, as IIUC, the on-chain execution in execution actors is performed independently in each subnet, so validators are trusting whatever external state the user trigger the execution is bringing from other subnets). At this point, my take would be to implement and write a clear spec with any of these options, and iterate from there. A lot of the security guarantees for the protocol currently relies on the implementation of the execution protocol, so it would be useful to start coming up with use cases and have something in place to make a more informed decision of the best approach and the security of the scheme. WDYT?

**Reply by sergefdrv** [reply]
@adlrocha Sorry for confusion, I should have summarized the difference. But let me first respond to your comment above.

> propagate the result up to the coordinator from the execution actor

If by results you mean the output state then no, the coordinator actor only receives concise identifiers.

> an additional RTT of cross-net messages from the execution actors and the coordinator to notify the output of the on-chain execution

Not really, there is no change in the number of cross-chain messages.

> the on-chain execution in execution actors is performed independently in each subnet

True

> so validators are trusting whatever external state the user trigger the execution is bringing from other subnets

Although, when running the execution actors in subnets, validators use inputs from other subnets supplied by users as is, they verify and irreversibly consume their own input and compute the atomic execution ID such that validators from all subnets must agree at least on the inputs (but potentially also on something else) in order to eventually commit the results. Therefore they rely on the correct execution of actors by _validators_ in the involved subnets. In the baseline, they don't assume correctness of off-chain execution results.

> implement and write a clear spec with any of these options

Do you mean to spec out a more concrete instantiation of the proposed scheme?

> it would be useful to start coming up with use cases

Do we already have clear examples of use-cases for the atomic execution?

**Reply by sergefdrv** [reply]
I added a remark to [the description](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477) according to [this](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3897165) comment.

**Reply by sergefdrv** [reply]
[The proposed version of the protocol](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477) is indeed similar to [this one](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3268374). It can be considered as a generalization with some tweaks. Nevertheless, I think there are some important differences:
 - It defaults to a more secure scheme because the protocol instantiation with the output computed by subnet validators on-chain is more straightforward to implement, whereas the alternative version with the output computed off-chain remains possible. This should help actors to ensure correctness of inputs/outputs and state updates.
 - It explicitly requires that inputs/locks must be unique so that independent atomic executions are guaranteed to get distinct execution IDs from their sets of inputs, even if the inputs refer to equivalent initial pieces of state.
 - It clearly specifies what the atomic execution ID should at least depend on.
 - Locking can be optionally/partially deferred until the pre-commit step, e.g. it enables optimistic locking schemes.
 - Last, but not least, I think it should cause less confusion because it explicitly captures the essence of the atomic execution protocol and its core requirements leaving the details for a particular instantiation. E.g., the notion of output would not be re-purposed in case we want to instantiate the scheme in a different way, c.f. [here](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3903691). It doesn't reserve particular method numbers or actor IDs.
 - Its extra flexibility should enable wider range of potential use cases.

**Reply by adlrocha** [reply]
I get your point, but is hard to clearly see the nuances between the two proposals without code (you know, the eternal fight we've always suffered between papers saying that something "is trivial to implement" and the implementation ending up being a complete nightmare). For instance, I would claim that the "optimistic locking" is not as simple to implement and securely handle as you suggest (in term of races and messages trying to access to state in parallel), but maybe I am wrong.

I think both proposals achieve the goal we are seeking with this protocol (in slightly different ways and with slightly different security/performance trade-offs). My suggestion would be to get one of them implemented ASAP (the one that you are more fond of, or easier to implement) and iterate the design over the code. Both of them seem similar enough to, once we have the basic architecture of the protocol, be able to tweak to our needs without major changes (but that way we prevent the "paralysis by analysis":) ). WDYT?

**Reply by sergefdrv** [reply]
> I would claim that the "optimistic locking" is not as simple to implement and securely handle

True, I'd also prefer starting with a conservative scheme, but have the protocol that explicitly allows the optimistic scheme to be implemented later.

> get one of them implemented ASAP (the one that you are more fond of, or easier to implement) and iterate the design over the code

Cool, I'm going over the code of [this PR](https://github.com/consensus-shipyard/builtin-actors/pull/17) and will take it from there.

**Reply by sergefdrv** [reply]
In order to simplify the implementation and to optimize the happy path, I'm implementing the scheme with the following changes:
 - removing `rollback` parameter from the `pre_commit` method of the coordinator actor and not storing it in the state
 - aborting at each execution actor individually (through the coordinator actor)
